### PR TITLE
[FSDP] Parameter shape verification and synchronization

### PR DESCRIPTION
**Pull request by rohan-varma**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #75977
* __->__ #75976
* #75970
* #75945

Use generalized APIs to perform param shape verification and module
synchronization.

Differential Revision: [D35712026](https://our.internmc.facebook.com/intern/diff/D35712026/)

### CODE REVIEW

File: torch/distributed/fsdp/fully_sharded_data_parallel.py
Comment (awgu) [inline torch/distributed/fsdp/fully_sharded_data_parallel.py:476]: nit: `to to` -> `to`

`device_ids` -> `device_id`? (We want to enforce SPSD for FSDP?)

Comment (awgu) [inline torch/distributed/fsdp/fully_sharded_data_parallel.py:476]: Is `self.compute_device` guaranteed to be a CUDA device (and not CPU)?

File: torch/distributed/fsdp/fully_sharded_data_parallel.py
Comment (awgu) [inline torch/distributed/fsdp/fully_sharded_data_parallel.py:481]: My understanding is that shared parameters will be verified and sync-ed multiple times, but there is not an easy way around this. Is this correct?

**Review by awgu — APPROVED** [review]
Stamping to unblock since the logic all makes sense to me. Left some minor comments.

File: torch/distributed/fsdp/fully_sharded_data_parallel.py
Comment (mrshenli) [reply torch/distributed/fsdp/fully_sharded_data_parallel.py:476]: line too long, shall we break it into shorter ones?

Comment (mrshenli) [reply torch/distributed/fsdp/fully_sharded_data_parallel.py:476]: Is it OK to mention `device_id` at this point, given that this arg is not available yet?

Comment (rohan-varma) [reply torch/distributed/fsdp/fully_sharded_data_parallel.py:476]: Was planning to base the next PR under this one and land this after that one lands and we agree on CPU-side semantics. Currently PR is mostly for discussion

File: torch/distributed/fsdp/fully_sharded_data_parallel.py
Comment (zhaojuanmao) [inline torch/distributed/fsdp/fully_sharded_data_parallel.py:492]: Could we move this to be after "self._shard_parameters()", so that we can flatten and shard parameters on GPU, this will greatly improve model init time when users pass a CPU model without device_id specified?

Comment (rohan-varma) [reply torch/distributed/fsdp/fully_sharded_data_parallel.py:492]: This is taken care of if users pass in `device_id`s already. Also this PR and `was_cpu` logic is mostly intended for discussion, as this does not seem like a great way to support CPU use cases?

Comment (zhaojuanmao) [reply torch/distributed/fsdp/fully_sharded_data_parallel.py:492]: that sounds good we only support GPU sharding when users pass a device id. although we can support GPU sharding and move sharded module from GPU to CPU even users do not pass a device id? but probably it is not a good user pattern...

**Comment by zhaojuanmao** [comment]
sorry, we need to benchmark this PR before landing

File: torch/distributed/fsdp/fully_sharded_data_parallel.py
Comment (rohan-varma) [reply torch/distributed/fsdp/fully_sharded_data_parallel.py:479]: Will remove it for now, and update warning when device_ids PR lands.

File: torch/distributed/fsdp/fully_sharded_data_parallel.py
Comment (rohan-varma) [reply torch/distributed/fsdp/fully_sharded_data_parallel.py:484]: Yeah, hopefully there are not too many shared params where it becomes a perf issue.

File: torch/distributed/fsdp/fully_sharded_data_parallel.py
Comment (rohan-varma) [reply torch/distributed/fsdp/fully_sharded_data_parallel.py:495]: Sure, per our discussion we can offer basic support for this and log a warning saying device_ids should be used. Will add warning in the next diff for device_Ids.

For this PR, if CPU, we'll move to compute_device (in the future compute_device == device_id should be an invariant), and run sharding, flattening as well on GPU before moving the module back to CPU.

File: torch/distributed/fsdp/fully_sharded_data_parallel.py
Comment (rohan-varma) [reply torch/distributed/fsdp/fully_sharded_data_parallel.py:479]: @awgu Yep it is

File: torch/distributed/fsdp/fully_sharded_data_parallel.py
Comment (rohan-varma) [reply torch/distributed/fsdp/fully_sharded_data_parallel.py:495]: We also call `del module` after sharding and before flattening, I think given this, if original model was on CPU, we can simply move it back similar to how we offload the param to CPU in ctor.

**Comment by rohan-varma** [comment]
Close in favor of https://github.com/pytorch/pytorch/pull/76442/files

**Pull request closed without merging**
